test/src/test/java/org/corfudb/runtime/checkpoint/CheckpointSmokeTest.java

```diff
@@ -456,6 +462,293 @@ public void checkpointWriterInterleavedTest() throws Exception {
+        } finally {
+            r.getObjectsView().TXEnd();
+        }
+    }
+
+    /**
+     * When log entry compression is set to NONE and BATCH_THRESHOLD_PERCENTAGE is very small,
+     * although the total size of SMR entries could pass the check in appendObjectState(), it
+     * could fail in writing the CheckpointEntry into stream (WriteSizeException) due to the
+     * slight overhead of packing SMR entries into CheckpointEntry.
+     */
+    @Test(expected = WriteSizeException.class)
+    @SuppressWarnings("checkstyle:magicnumber")
+    public void checkpointWriterSizeLimitViolationTest() throws Exception {
+        final String streamName = "mystream7";
+        final UUID streamId = CorfuRuntime.getStreamID(streamName);
+        final String keyPrefix = "a-prefix";
+        final String author = "Me, myself, and I";
+        // max write size set to 25 MB
+        final int maxWriteSize = 25 << 20;
+
+        StreamingMap<String, String> m = instantiateStringMap(streamName);
+
+        // set the max write size to 25 MB
+        getRuntime().getParameters().setMaxWriteSize(maxWriteSize);
+        getRuntime().getParameters().setCodecType(Codec.Type.NONE);
+        CheckpointWriter cpw = new CheckpointWriter(getRuntime(), streamId, author, (CorfuTable) m);
+        cpw.setSerializer(serializer);
+        final double batchThresholdPercentage = 1;
+        cpw.setBATCH_THRESHOLD_PERCENTAGE(batchThresholdPercentage);
+        cpw.setBatchSize(Integer.MAX_VALUE);
+
+        int i = 0;
+        int numBytesPerCheckpointEntry = 0;
+        // first entry is 20 MB
+        int payloadLength = 20 << 20;
+        String payload = getRandomStringOfSize(payloadLength);
+        int nextSmrEntrySize = getSerializedSMREntrySize(keyPrefix + i, payload,
+                cpw.getKeyMutator(), cpw.getValueMutator(), cpw.getSerializer());
+        int reducedMaxWriteSize = (int)Math.floor(batchThresholdPercentage * maxWriteSize);
+
+        // Create entries whose size add up to be in range [reducedMaxWriteSize - 100, reducedMaxWriteSize)
+        while (numBytesPerCheckpointEntry + 100 < reducedMaxWriteSize) {
+            if (nextSmrEntrySize + numBytesPerCheckpointEntry >= reducedMaxWriteSize) {
+                payloadLength /= 2;
+                payload = getRandomStringOfSize(payloadLength);
+                nextSmrEntrySize = getSerializedSMREntrySize(keyPrefix + i, payload,
+                        cpw.getKeyMutator(), cpw.getValueMutator(), cpw.getSerializer());
+                continue;
+            }
+            m.put(keyPrefix + i, payload);
+            numBytesPerCheckpointEntry += nextSmrEntrySize;
+            System.out.println(String.format("After putting entry %s (payload len is %s, size is %s), numBytesPerCheckpointEntry is: %s",
+                    (keyPrefix + i), payloadLength, nextSmrEntrySize, numBytesPerCheckpointEntry));
+            System.out.println("-------------------------------------------------------------");
```

Comment:
Logging shouldn't be used in tests, and especially `System.out.`. Instead, these should be asserts.

Reply:
@Lujie1996 should I remove this out?

Reply:
I think it's already removed in my most recent commit https://github.com/CorfuDB/CorfuDB/pull/2875/commits/d02a4512ea51298b833370593ee9c23578171705